File: src/test/java/de/fraunhofer/aisec/codyze/crymlin/CommandsTest.java

package de.fraunhofer.aisec.codyze.crymlin;

import de.fraunhofer.aisec.codyze.Commands;
import de.fraunhofer.aisec.codyze.JythonInterpreter;
import de.fraunhofer.aisec.codyze.analysis.AnalysisServer;
import de.fraunhofer.aisec.codyze.analysis.ServerConfiguration;
import de.fraunhofer.aisec.codyze.analysis.TypestateMode;
import org.junit.jupiter.api.Test;

import java.io.*;

import static org.junit.jupiter.api.Assertions.*;

/** Testing console commands. */
class CommandsTest {

	@Test
	void commandConsoleTest() {
		AnalysisServer server = AnalysisServer.builder()
				.config(ServerConfiguration.builder()
						.launchLsp(false)
						.launchConsole(true)
						.typestateAnalysis(TypestateMode.NFA)
						.markFiles(new File("src/test/resources/mark_java").getAbsolutePath())
						.useLegacyEvaluator()
						.build())
				.build();

		server.start();
		PrintStream oldOut = System.out;
		PrintStream oldErr = System.err;
		ByteArrayOutputStream bosOut = new ByteArrayOutputStream();
		PrintStream o = new PrintStream(bosOut);
		ByteArrayOutputStream bosErr = new ByteArrayOutputStream();
		PrintStream e = new PrintStream(bosErr);

		System.setOut(o);
		System.setErr(e);

		// Init console
		JythonInterpreter interp = new JythonInterpreter();
		Commands com = new Commands(interp);

		Commands.help();
		assertTrue(bosOut.toString().contains("help"));

		com.load_rules("src/test/resources/mark_java");

		com.list_rules();
		assertTrue(bosOut.toString().contains("Foo"));

		com.analyze("src/test/resources/good/Bouncycastle.java");
		com.show_findings();

		System.setOut(oldOut);
		System.setErr(oldErr);
	}
}

File: src/test/java/de/fraunhofer/aisec/codyze/crymlin/CommandsTest.kt
package de.fraunhofer.aisec.codyze.crymlin

import de.fraunhofer.aisec.codyze.Commands
import de.fraunhofer.aisec.codyze.JythonInterpreter
import de.fraunhofer.aisec.codyze.analysis.AnalysisServer
import de.fraunhofer.aisec.codyze.analysis.TypestateMode
import de.fraunhofer.aisec.codyze.config.Configuration
import java.io.*
import kotlin.test.assertTrue
import org.junit.jupiter.api.*

/** Testing console commands. */
internal class CommandsTest {
    @Test
    fun commandConsoleTest() {
        val config = Configuration()
        config.codyze.executionMode.isLsp = false
        config.codyze.executionMode.isCli = true
        config.codyze.analysis.tsMode = TypestateMode.DFA
        config.codyze.mark = arrayOf(File("src/test/resources/mark_java"))

        val server = AnalysisServer.builder().config(config).build()
        server.start()
        val oldOut = System.out
        val oldErr = System.err
        val bosOut = ByteArrayOutputStream()
        val o = PrintStream(bosOut)
        val bosErr = ByteArrayOutputStream()
        val e = PrintStream(bosErr)
        System.setOut(o)
        System.setErr(e)

        // Init console
        val interp = JythonInterpreter()
        val com = Commands(interp)
        Commands.help()
        assertTrue(bosOut.toString().contains("help"))

        com.load_rules("src/test/resources/mark_java")
        com.list_rules()
        assertTrue(bosOut.toString().contains("Foo"))

        com.analyze("src/test/resources/good/Bouncycastle.java")
        com.show_findings()
        System.setOut(oldOut)
        System.setErr(oldErr)
    }
}
